### PR TITLE
Bug 1111524 - Add safety timeout for missing backdrop event +autoland

### PR DESCRIPTION
**Pull request by KevinGrandon**
https://bugzilla.mozilla.org/show_bug.cgi?id=1111524

Also create a small shared library to help guard event timeouts.

### CODE REVIEW

**Comment by try-server-hook** [comment]
[KevinGrandon Kevin Grandon (KevinGrandon)](https://github.com/KevinGrandon) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=23c95f79838d)

**Comment by try-server-hook** [comment]
[KevinGrandon Kevin Grandon (KevinGrandon)](https://github.com/KevinGrandon) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=6b336e0aef8f)

File: apps/sharedtest/test/unit/event_safety_test.js [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+suite('eventSafety', function() {
+
+  test('fires after event', function(done) {
+    this.sinon.useFakeTimers();
+    eventSafety(window, 'foobar', done, 1000);
+    window.dispatchEvent(new CustomEvent('foobar'));
+
+    // Tick the clock to ensure that the timeout is cleared.
+    // Sinon will throw an error if done is called multiple times.
+    this.sinon.clock.tick(1000);
```
Comment (etiennesegonzac) [inline apps/sharedtest/test/unit/event_safety_test.js]: can we move those into a teardown, I feel like it makes the test much more expressive

File: apps/sharedtest/test/unit/event_safety_test.js [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+  });
+
+  test('fires after setTimeout', function(done) {
+    this.sinon.useFakeTimers();
+    eventSafety(window, 'foobar', done, 1000);
+    this.sinon.clock.tick(1000);
+
+    // Fire an event to ensure that the event is unsubscribed from.
+    // Sinon will throw an error if done is called multiple times.
+    window.dispatchEvent(new CustomEvent('foobar'));
```
Comment (etiennesegonzac) [inline apps/sharedtest/test/unit/event_safety_test.js]: same

File: apps/sharedtest/test/unit/event_safety_test.js [hunk 3]
```diff
@@ -0,0 +1,28 @@
+'use strict';
+/* global eventSafety */
+
+require('/shared/js/event_safety.js');
+
+suite('eventSafety', function() {
+
+  test('fires after event', function(done) {
+    this.sinon.useFakeTimers();
+    eventSafety(window, 'foobar', done, 1000);
```
Comment (etiennesegonzac) [inline apps/sharedtest/test/unit/event_safety_test.js]: nit: 
```
function() {
  assert.ok(true, 'callback fired by event');
  done();
}
```

File: apps/sharedtest/test/unit/event_safety_test.js [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+    window.dispatchEvent(new CustomEvent('foobar'));
+
+    // Tick the clock to ensure that the timeout is cleared.
+    // Sinon will throw an error if done is called multiple times.
+    this.sinon.clock.tick(1000);
+  });
+
+  test('fires after setTimeout', function(done) {
+    this.sinon.useFakeTimers();
+    eventSafety(window, 'foobar', done, 1000);
```
Comment (etiennesegonzac) [inline apps/sharedtest/test/unit/event_safety_test.js]: same little assert message

File: apps/sharedtest/test/unit/event_safety_test.js (hunk 3 above)
Comment (KevinGrandon) [inline apps/sharedtest/test/unit/event_safety_test.js]: The problem here is that asserting on ```true``` will never trigger the message. It makes it show up in the success case, but I'm not sure how valuable that is. I'll go ahead and add it regardless though to get this landed.

**Comment by try-server-hook** [comment]
[KevinGrandon Kevin Grandon (KevinGrandon)](https://github.com/KevinGrandon) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=6528a99c73ef)

**Comment by try-server-hook** [comment]
[KevinGrandon Kevin Grandon (KevinGrandon)](https://github.com/KevinGrandon) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=78d438510483)

**Pull request merged**